moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,1260 @@
+                                              shared_variables_.command_deltas.twist.linear.z == 0.0 &&
+                                              shared_variables_.command_deltas.twist.angular.x == 0.0 &&
+                                              shared_variables_.command_deltas.twist.angular.y == 0.0 &&
+                                              shared_variables_.command_deltas.twist.angular.z == 0.0;
+  pthread_mutex_unlock(&shared_variables_.zero_cartesian_cmd_flag_mutex);
+
+  pthread_mutex_unlock(&shared_variables_.command_deltas_mutex);
+
+  pthread_mutex_lock(&shared_variables_.incoming_cmd_stamp_mutex);
+  shared_variables_.incoming_cmd_stamp = msg->header.stamp;
+  pthread_mutex_unlock(&shared_variables_.incoming_cmd_stamp_mutex);
+}
+
+// Listen to joint delta commands.
+// Store them in a shared variable.
+void JogROSInterface::deltaJointCmdCB(const moveit_experimental::JogJointConstPtr& msg)
+{
+  pthread_mutex_lock(&shared_variables_.joint_command_deltas_mutex);
+  shared_variables_.joint_command_deltas = *msg;
+
+  // Input frame determined by YAML file
+  shared_variables_.joint_command_deltas.header.frame_id = ros_parameters_.command_frame;
+
+  // Check if joint inputs is all zeros. Flag it if so to skip
+  // calculations/publication
+  bool all_zeros = true;
+  for (double delta : shared_variables_.joint_command_deltas.deltas)
+  {
+    all_zeros &= (delta == 0.0);
+  };
+  pthread_mutex_unlock(&shared_variables_.joint_command_deltas_mutex);
+
+  pthread_mutex_lock(&shared_variables_.zero_joint_cmd_flag_mutex);
+  shared_variables_.zero_joint_cmd_flag = all_zeros;
+  pthread_mutex_unlock(&shared_variables_.zero_joint_cmd_flag_mutex);
+
+  pthread_mutex_lock(&shared_variables_.incoming_cmd_stamp_mutex);
+  shared_variables_.incoming_cmd_stamp = msg->header.stamp;
+  pthread_mutex_unlock(&shared_variables_.incoming_cmd_stamp_mutex);
+}
+
+// Listen to joint angles.
+// Store them in a shared variable.
+void JogROSInterface::jointsCB(const sensor_msgs::JointStateConstPtr& msg)
+{
+  pthread_mutex_lock(&shared_variables_.joints_mutex);
+  shared_variables_.joints = *msg;
+  pthread_mutex_unlock(&shared_variables_.joints_mutex);
+}
+
+// Read ROS parameters, typically from YAML file
+bool JogROSInterface::readParameters(ros::NodeHandle& n)
+{
+  std::size_t error = 0;
+
+  // Specified in the launch file. All other parameters will be read
+  // from this namespace.
+  std::string parameter_ns;
+  ros::param::get("~parameter_ns", parameter_ns);
+  if (parameter_ns == "")
+  {
+    ROS_ERROR_STREAM_NAMED(NODE_NAME, "A namespace must be specified in the launch file, like:");
+    ROS_ERROR_STREAM_NAMED(NODE_NAME, "<param name=\"parameter_ns\" "
+                                      "type=\"string\" "
+                                      "value=\"left_jog_arm_server\" />");
+    return 0;
+  }
+
+  error += !rosparam_shortcuts::get("", n, parameter_ns + "/publish_period", ros_parameters_.publish_period);
```

Comment:
Looks like this `rosparam_shortcuts` thing is somethign @davetcoleman made? In this particular case, I think the code would be just as readable (or more so) if you created a namespaced node handle and then used it to read the params rather than calling a utility function. No change required though.

Reply:
One nice thing about rosparam_shortcuts is it prints an error message then exits if there's a missing parameter.

Reply:
Is it released in kinetic and melodic? Or is this something we should add to the `moveit.rosinstall` in the root of the repo?

Reply:
Its been released in every ROS distro since Indigo and I release it very quick compared to e.g. moveit ;-)